### PR TITLE
    Perception: major change in motion service/lane tracking. pass stress test.

### PR DESCRIPTION
**Pull request by gchen-apollo**
add different strategies to handle lane tracking, with callback on motion: (1)motion arrives earlier (2) motion later (3) motion not found. (4)motion drop (5) image drop. etc

### CODE REVIEW

File: modules/perception/obstacle/onboard/lane_post_processing_subnode.cc [hunk 1]
```diff
@@ -234,17 +235,78 @@ Status LanePostProcessingSubnode::ProcEvents() {
...
+      }
+    } else {
+      mutex_.lock();
+      options_.SetMotion(motion_service_->GetMotionBuffer()->back());
+      mutex_.unlock();
+    }
     AINFO << "options_.vehicle_status.motion:  "
           << options_.vehicle_status.motion;
+
+    // while (motion_buffer.empty() ||
```
Comment (ycool) [inline modules/perception/obstacle/onboard/lane_post_processing_subnode.cc]: why keep comment out codes ?

File: modules/perception/obstacle/onboard/motion_service.cc [hunk 2]
```diff
@@ -124,12 +138,19 @@ void MotionService::OnLocalization(
...
       PublishEvent(camera_timestamp);
     } else {
-      AERROR << "camera timestamp should arrive in order";
-      return;
+      ADEBUG << "Motion_status: pop";
+      vehicle_planemotion_->add_new_motion(
+          &vehicle_status, pre_camera_timestamp_, camera_timestamp,
+          PlaneMotion::RESET);
+
+      // AERROR << "camera timestamp should arrive in order";
```
Comment (ycool) [inline modules/perception/obstacle/onboard/motion_service.cc]: remove comment codes

File: modules/perception/onboard/common_shared_data.h [hunk 3]
```diff
@@ -202,8 +202,8 @@ template <class M>
 bool CommonSharedData<M>::Add(const CommonSharedDataKey &key,
                               const SharedDataPtr<M> &data) {
   // update latest_timestamp for SharedData
-  latest_timestamp_ = std::max(latest_timestamp_, key.timestamp);
-  // latest_timestamp_ = key.timestamp;
+  // latest_timestamp_ = std::max(latest_timestamp_, key.timestamp);
```
Comment (ycool) [inline modules/perception/onboard/common_shared_data.h]: remove comment code

File: modules/perception/obstacle/camera/lane_post_process/cc_lane_post_processor/cc_lane_post_processor.cc [hunk 4]
```diff
@@ -829,18 +829,13 @@ bool CCLanePostProcessor::Process(const cv::Mat &lane_map,
       }
     }
 
-//    AINFO << "History buffer size: " << lane_history_.size();
     for (l = 0; l < is_valid.size(); l++) {
```
[the remark this answers is not in the document]
Comment (gchen-apollo) [reply modules/perception/obstacle/camera/lane_post_process/cc_lane_post_processor/cc_lane_post_processor.cc]: the l uses as a out variable as well. not able to define inside. But I change the style.

File: modules/perception/obstacle/onboard/motion_service.cc [hunk 5]
```diff
@@ -124,36 +134,17 @@ void MotionService::OnLocalization(
...
-  //   }
-
-  //  AINFO << "Motion Matrix: ";
-  //  auto motion_buffer_ptr = vehicle_planemotion_->get_buffer();
-  //  int motion_size = motion_buffer_ptr->size();
-  //  AINFO << (*motion_buffer_ptr)[motion_size-1].motion;
-  //  AINFO << "Motion Matrix end";
+
+  {
+    MutexLock lock(&image_mutex_);
```
[the remark this answers is not in the document]
Comment (gchen-apollo) [reply modules/perception/obstacle/onboard/motion_service.cc]: same thing. MutexLock is like std::lock_guard<std::mutex>

**Pull request merged**